src/test/key_tests.cpp

```diff
@@ -345,4 +346,31 @@ BOOST_AUTO_TEST_CASE(bip341_test_h)
     BOOST_CHECK(XOnlyPubKey::NUMS_H == H);
 }
 
+BOOST_AUTO_TEST_CASE(key_schnorr_tweak_smoke_test)
+{
+    // Sanity check to ensure we get the same tweak using CPubKey vs secp256k1 functions
+    secp256k1_context* secp256k1_context_sign = secp256k1_context_create(SECP256K1_CONTEXT_SIGN);
+
+    CKey key;
+    key.MakeNewKey(true);
+    uint256 merkle_root = InsecureRand256();
+
+    // secp256k1 functions
+    secp256k1_keypair keypair;
+    BOOST_CHECK(secp256k1_keypair_create(secp256k1_context_sign, &keypair, UCharCast(key.begin())));
+    secp256k1_xonly_pubkey pubkey_old;
+    BOOST_CHECK(secp256k1_keypair_xonly_pub(secp256k1_context_sign, &pubkey_old, nullptr, &keypair));
+    unsigned char pubkey_bytes_old[32];
+    BOOST_CHECK(secp256k1_xonly_pubkey_serialize(secp256k1_context_sign, pubkey_bytes_old, &pubkey_old));
+    uint256 tweak_old = XOnlyPubKey(pubkey_bytes_old).ComputeTapTweakHash(merkle_root.IsNull() ? nullptr : &merkle_root);
```

Comment:
This is how @paplorinc originally proposed the test: https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1685696117

I felt like computing multiple keys in a loop was overkill and removed the merkle root switching since that's not really what's being tested here. The goal here is to sanity check that we get the same public key bytes from GetPubKey() as we would get directly accessing the `secp256k1_keypair` using the `_xonly_pub` and `_serialize`.

So perhaps better to just remove the `IsNull()` check, considering this is being checked in the manner you describe in the `bip340_test_vectors` test?

Reply:
I've updated the to remove the `merkle_root.IsNull()` check.

@ismaelsadeeq I kept the `ComputeTapTweakHash` line because they are testing _slightly_ different paths: in the first case, an `XOnlyPubKey` is being constructed from a 32 byte unsigned char array, in the second case it's being created from a `CPubKey` object. I renamed the variables to try to make this more clear in the test.

Reply:
Agree the code paths are very similar but I think the same argument as https://github.com/bitcoin/bitcoin/pull/30051#discussion_r1702771127 applies here (admittedly much weaker).